### PR TITLE
RavenDB-7308 & RavenDB-7423

### PR DESCRIPTION
**Pull request by karmeli87**

### CODE REVIEW

**Comment by karmeli87** [comment]
Still need to check how does those changes affect the studio.

File: src/Raven.Client/Documents/Replication/ReplicationNode.cs [hunk 1]
```diff
@@ -121,21 +64,13 @@ public override int GetHashCode()
         {
             unchecked
             {
-                var hashCode = (NodeTag?.GetHashCode() ?? 0);
-                hashCode = (hashCode * 397) ^ (Database?.GetHashCode() ?? 0);
-                hashCode = (hashCode * 397) ^ (ClientVisibleUrl?.GetHashCode() ?? 0);
-                return hashCode;
+                var hashCode = CalculateStringHash(Url);
+                hashCode = (hashCode * 397) ^ CalculateStringHash(Database);
+                hashCode = (hashCode * 397) ^ CalculateStringHash(Disabled.ToString());
```
Comment (ayende) [inline src/Raven.Client/Documents/Replication/ReplicationNode.cs:69]: Why generate a string instance here?

File: src/Raven.Client/Server/DatabaseTopology.cs [hunk 2]
```diff
@@ -1,265 +1,171 @@
...
-        {
-            return
-                ResolveByCollection?.Count == 0 &&
-                ResolveToLatest == false &&
-                DatabaseResolverId == null;
-        }
+        ulong GetTaskKey();
+    }
+//
+//    public class DatabaseTopologyNode : ReplicationNode, IDatabaseTask, IComparable<DatabaseTopologyNode>
```
Comment (ayende) [inline src/Raven.Client/Server/DatabaseTopology.cs:16]: Why do we still have commented code here?

File: src/Raven.Client/Server/DatabaseTopology.cs [hunk 3]
```diff
@@ -1,265 +1,171 @@
...
         }
 
-        public override bool Equals(object obj)
+        protected static ulong CalculateStringHash(string s)
         {
-            var resolver = obj as ScriptResolver;
-            if (resolver == null)
-                return false;
-            return string.Equals(Script, resolver.Script, StringComparison.OrdinalIgnoreCase) && LastModifiedTime == resolver.LastModifiedTime;
+            return string.IsNullOrEmpty(s) ? 0 : Hashing.XXHash64.Calculate(s, Encodings.Utf8);
```
Comment (ayende) [inline src/Raven.Client/Server/DatabaseTopology.cs:71]: What is wrong with using the string hash here?

File: src/Raven.Client/Server/ExternalReplication.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+using Sparrow.Json.Parsing;
+
+namespace Raven.Client.Server
+{
+    public class ExternalReplication : ReplicationNode, IDatabaseTask, IDynamicJsonValueConvertible
+    {
+        public string ApiKey;
+        public long TaskId;
+
+        public static void RemoveWatcher(ref List<ExternalReplication> watchers, long taskId)
```
Comment (ayende) [inline src/Raven.Client/Server/ExternalReplication.cs:16]: Why is this ref?

File: src/Raven.Client/Server/ExternalReplication.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+        public static void EnsureUniqueDbAndUrl(List<ExternalReplication> watchers, ExternalReplication watcher)
+        {
+            var dbName = watcher.Database;
+            var url = watcher.Url;
+            foreach (var w in watchers)
+            {
+                if (w.Database != dbName || w.Url != url)
+                    continue;
+                watchers.Remove(watcher);
```
Comment (ayende) [inline src/Raven.Client/Server/ExternalReplication.cs:35]: Can it never be the same instance here?

File: test/FastTests/Issues/RavenDB-6173.cs [hunk 6]
```diff
@@ -64,30 +64,30 @@ public class Bar
...
-
-                Assert.Equal(idOfD, nodesOfA.Outgoing.FirstOrDefault(x => x.DbId == idOfD)?.DbId);
-                Assert.Equal(false, nodesOfA.Outgoing.FirstOrDefault(x => x.DbId == idOfD).IsETLNode);
-
-                Assert.Equal(1, nodesOfB.Outgoing.Count);
-                Assert.Equal(idOfC, nodesOfB.Outgoing.FirstOrDefault(x => x.DbId == idOfC)?.DbId);
-                Assert.Equal(true, nodesOfB.Outgoing.FirstOrDefault(x => x.DbId == idOfC)?.IsETLNode);
-
-                Assert.Equal(0, nodesOfC.Outgoing.Count);
+//                var topologyInfo = GetLiveTopology(storeA);
```
Comment (ayende) [inline test/FastTests/Issues/RavenDB-6173.cs:67]: Why is this commented?

**Pull request merged**